README.md
# Spiderchart

An awesome radar charts to make your app just awesome

![Radar chart flutter example](https://github.com/g-apparence/spiderchart/blob/master/example/assets/spiderchart.gif)

## Getting Started
Creating a SpiderChart is that easy
you can choose the color of the lines, radius of point. 
A SpiderChartValue is all the stats for the group you want to compare with your spider or radar graph. 

```dart
SpiderChartWidget component = SpiderChartWidget(
  lineColor: Colors.blueGrey[500],
  activeLabelColor: Colors.blueGrey[50],
  nbSides: 7,
  labels: [
    SpiderChartStatModel(roundSize: 4, label: "test 1"),
    SpiderChartStatModel(roundSize: 4, label: "test 2"),
    SpiderChartStatModel(roundSize: 4, label: "test 3"),
    SpiderChartStatModel(roundSize: 4, label: "test 4"),
    SpiderChartStatModel(roundSize: 4, label: "test 5"),
    SpiderChartStatModel(roundSize: 4, label: "test 6"),
    SpiderChartStatModel(roundSize: 4, label: "test 7"),
  ],
  valuesList: new List()
    ..add(
      new SpiderChartValue(
        [ .5, .8, .9, .8, .9, .1, .2 ],
        gradient: gradient,
        strokeColor: Colors.lightGreenAccent[200]
      )
    ),
);
```

Enjoy


CHANGELOG.md
## [0.0.1] - 02/06/2020
* First version of the spider chart library


test/spiderchart_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:awesome_spiderchart/chart_painter.dart';

import 'package:awesome_spiderchart/spiderchart.dart';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';
import 'screen_size_extension.dart';

void main() {
  group('spiderchart with 5 labels with gradient', () {

    final Gradient gradient = new LinearGradient(
      colors: <Color>[
        Colors.lightGreenAccent[100].withOpacity(0.7),
        Colors.lightGreenAccent[700].withOpacity(0.7),
      ],
      begin: Alignment(1.0, -1.0),
      end: Alignment(-1.0, 1.0),
    );


    SpiderChartWidget component = SpiderChartWidget(
      lineColor: Colors.blueGrey[500],
      activeLabelColor: Colors.blueGrey[50],
      nbSides: 5,
      labels: [
        SpiderChartStatModel(roundSize: 4, label: "test 1"),
        SpiderChartStatModel(roundSize: 4, label: "test 2"),
        SpiderChartStatModel(roundSize: 4, label: "test 3"),
        SpiderChartStatModel(roundSize: 4, label: "test 4"),
        SpiderChartStatModel(roundSize: 4, label: "test 5"),
      ],
      valuesList: new List()
        ..add(
          new SpiderChartValue(
            [ .5, .8, .9, .8, .9 ],
            gradient: gradient,
            strokeColor: Colors.lightGreenAccent[200]
          )
        ),
    );

    testWidgets('widgets exists', (WidgetTester tester) async {
      var app = new MediaQuery(data: MediaQueryData(), child: MaterialApp(
        home: Scaffold(
          body: component,
        )
      ));
      await tester.pumpWidget(app);
      expect(find.byType(component.runtimeType), findsOneWidget);
    });

    testWidgets('[IPHONE 11] widgets must have this exact design', (WidgetTester tester) async {
      await tester.setIphone11Max();
      // create widget
      var app = MaterialApp(home: Scaffold(body: component));
      await tester.pumpWidget(app);
      // create golden file
      await expectLater(find.byType(component.runtimeType), matchesGoldenFile("goldens/iphone11_5_labels_chart.png"));
    });

    testWidgets('[IPHONE 8 plus] widgets must have this exact design', (WidgetTester tester) async {
      await tester.setIphone8Plus();
      // create widget
      var app = MaterialApp(home: Scaffold(body: component));
      await tester.pumpWidget(app);
      // create golden file
      await expectLater(find.byType(component.runtimeType), matchesGoldenFile("goldens/iphone8Plus_5_labels_chart.png"));
    });

  });


  group('spiderchart with 7 labels with gradient', () {
    final Gradient gradient = new LinearGradient(
      colors: <Color>[
        Colors.lightGreenAccent[100].withOpacity(0.7),
        Colors.lightGreenAccent[700].withOpacity(0.7),
      ],
      begin: Alignment(1.0, -1.0),
      end: Alignment(-1.0, 1.0),
    );


    SpiderChartWidget component = SpiderChartWidget(
      lineColor: Colors.blueGrey[500],
      activeLabelColor: Colors.blueGrey[50],
      nbSides: 7,
      labels: [
        SpiderChartStatModel(roundSize: 4, label: "test 1"),
        SpiderChartStatModel(roundSize: 4, label: "test 2"),
        SpiderChartStatModel(roundSize: 4, label: "test 3"),
        SpiderChartStatModel(roundSize: 4, label: "test 4"),
        SpiderChartStatModel(roundSize: 4, label: "test 5"),
        SpiderChartStatModel(roundSize: 4, label: "test 6"),
        SpiderChartStatModel(roundSize: 4, label: "test 7"),
      ],
      valuesList: new List()
        ..add(
          new SpiderChartValue(
            [ .5, .8, .9, .8, .9, .1, .2 ],
            gradient: gradient,
            strokeColor: Colors.lightGreenAccent[200]
          )
        ),
    );

    testWidgets('widgets exists', (WidgetTester tester) async {
      var app = new MediaQuery(data: MediaQueryData(), child: MaterialApp(
        home: Scaffold(
          body: component,
        )
      ));
      await tester.pumpWidget(app);
      expect(find.byType(component.runtimeType), findsOneWidget);
    });

    testWidgets('[IPHONE 11] widgets must have this exact design', (WidgetTester tester) async {
      await tester.setIphone11Max();
      // create widget
      var app = MaterialApp(home: Scaffold(body: component));
      await tester.pumpWidget(app);
      // create golden file
      await expectLater(find.byType(component.runtimeType), matchesGoldenFile("goldens/iphone11_7_labels_chart.png"));
    });

    testWidgets('[IPHONE 8 plus] widgets must have this exact design', (WidgetTester tester) async {
      await tester.setIphone8Plus();
      // create widget
      var app = MaterialApp(home: Scaffold(body: component));
      await tester.pumpWidget(app);
      // create golden file
      await expectLater(find.byType(component.runtimeType), matchesGoldenFile("goldens/iphone8Plus_7_labels_chart.png"));
    });

  });


  group('check bad configurations', () {

    testWidgets('creating a widget with less label than sides throws', (WidgetTester tester) async {
      final Gradient gradient = new LinearGradient(
        colors: <Color>[
          Colors.lightGreenAccent[100].withOpacity(0.7),
          Colors.lightGreenAccent[700].withOpacity(0.7),
        ],
      );
      var exception;
      try {
        SpiderChartWidget component = SpiderChartWidget(
          lineColor: Colors.blueGrey[500],
          activeLabelColor: Colors.blueGrey[50],
          nbSides: 12,
          labels: [
            SpiderChartStatModel(roundSize: 4, label: "test 1"),
            SpiderChartStatModel(roundSize: 4, label: "test 2"),
            SpiderChartStatModel(roundSize: 4, label: "test 3"),
            SpiderChartStatModel(roundSize: 4, label: "test 4"),
            SpiderChartStatModel(roundSize: 4, label: "test 5"),
            SpiderChartStatModel(roundSize: 4, label: "test 6"),
            SpiderChartStatModel(roundSize: 4, label: "test 7"),
          ],
          valuesList: new List()
            ..add(
              new SpiderChartValue(
                [ .5, .8, .9, .8, .9, .1, .2 ],
                gradient: gradient,
                strokeColor: Colors.lightGreenAccent[200]
              )
            ),
        );
        var app = MaterialApp(home: Scaffold(body: component));
        await tester.pumpWidget(app);
      } catch (e) {
        exception = e;
      }
      expect(exception , isNotNull);
    });

  });
}



test/screen_size_extension.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

extension ScreenSizeManager on WidgetTester {

  Future<void> setScreenSize({double width = 540, double height = 960, double pixelDensity = 1}) async {
    final size = Size(width, height);
    await this.binding.setSurfaceSize(size);
    this.binding.window.physicalSizeTestValue = size;
    this.binding.window.devicePixelRatioTestValue = pixelDensity;
  }

  // works for Iphone 11 max
  Future<void> setIphone11Max() => this.setScreenSize(width: 414, height: 896, pixelDensity: 3);

  // works for iphones size : 6+, 6s, 7+, 8+
  Future<void> setIphone8Plus() => this.setScreenSize(width: 414, height: 736, pixelDensity: 3);


}

lib/spiderchart.dart
library spiderchart;

import 'package:flutter/material.dart';
import 'chart_painter.dart';
import 'dart:ui' as ui;

import 'label_widget.dart';
import 'pointer_widget.dart';

class SpiderChartStatModel {
  ui.Image image;
  double roundSize;
  String label;

  SpiderChartStatModel({this.image, this.roundSize, this.label});
}



typedef OnTapLabel(int index);


class SpiderChartWidget extends StatefulWidget {

  final Color lineColor, activeLabelColor;
  final List<SpiderChartStatModel> labels;
  final List<SpiderChartValue> valuesList;
  final int nbSides;
  final OnTapLabel onTap;
  final int activeLabel;

  SpiderChartWidget({Key key, this.lineColor, this.labels, this.valuesList, this.nbSides, this.onTap, this.activeLabel = -1, this.activeLabelColor})
    : super(key: key) {
    assert(nbSides <= labels.length, "a label for each side must be provided ($nbSides, ${labels.length})");
  }

  @override
  _SpiderChartWidgetState createState() => _SpiderChartWidgetState();
}

class _SpiderChartWidgetState extends State<SpiderChartWidget> with SingleTickerProviderStateMixin {

  AnimationController _labelController;
  Animation<double> _scaleAnim;

  @override
  void initState() {
    super.initState();
    _labelController  = AnimationController(vsync: this, duration: new Duration(milliseconds: 1000));
    _scaleAnim = new CurvedAnimation(parent: _labelController, curve: Curves.easeInOut);
    _labelController.repeat(reverse: true);
  }

  @override
  void dispose() {
    _labelController.dispose();
    super.dispose();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        var width = constraints.maxWidth;
        var height = constraints.maxHeight;
        return ConstrainedBox(
          constraints: BoxConstraints(
            minWidth: width,
            minHeight: height,
            maxWidth: width,
            maxHeight: height
          ),
          child: Stack(
            children: new List()
              ..add(buildChart())
              ..addAll(List.generate(widget.nbSides, (final int index) => _createLabel(index))),
          ),
        );
      },
    );
  }

  Widget _createLabel(final int i) {
    return Positioned.fill(
      child: GestureDetector(
        key: ValueKey("tap$i"),
        onTap: () {
          if (widget.onTap != null) {
            widget.onTap(i);
          }
        },
        child: AnimatedBuilder(
          animation: _scaleAnim,
          builder: (context, child) {
            return widget.activeLabel == i ? PointerWidget(
              sideIndex: widget.activeLabel,
              nbSides: widget.nbSides,
              color: Colors.white,
              gradient: new RadialGradient(
                radius: .5,
                colors: <Color>[
                  Colors.lightGreenAccent.withOpacity(0),
                  widget.activeLabelColor,
                ],
              ),
              radius: widget.labels[i].roundSize * 4 + (widget.labels[i].roundSize * 2 * _scaleAnim.value),
            ) : child;
          },
          child: LabelWidget(
            index: i,
            lineColor: widget.lineColor,
            nbSides: widget.nbSides,
            statModel: widget.labels[i],
          ),
        ),
      ),
    );
  }

  Widget buildChart() {
    return Positioned.fill(
      child: CustomPaint(
        painter: SpiderChartDrawer(
          lineColor: widget.lineColor,
          valuesList: widget.valuesList,
          nbSides: widget.nbSides,
        ),
      ),
    );
  }

}










lib/chart_painter.dart
import 'dart:math';
import 'dart:ui' as ui;
import 'offset_extension.dart';
import 'package:flutter/material.dart';



class SpiderChartValue {
  /// [O..1] percentage of line
  List<double> values;

  Color bgColor, strokeColor;

  final Gradient gradient;

  SpiderChartValue(this.values, {this.bgColor, this.strokeColor, this.gradient});
}


class SpiderChartDrawer extends CustomPainter {

  final Color lineColor;

  final Color activeBgColor = Colors.lightGreenAccent[200];

  final int nbSides;

  final List<SpiderChartValue> valuesList;

  SpiderChartDrawer({
    this.lineColor,
    this.nbSides = 8,
    this.valuesList,
  });

  @override
  void paint(Canvas canvas, Size size) {
    var linePainter = Paint()
      ..style = PaintingStyle.stroke
      ..color = this.lineColor
      ..strokeWidth = 2
      ..isAntiAlias = true;
    // draw spider
    var center = Offset(size.width / 2, size.height / 2);
    var paneSize = size.width / 2 - 32;
    canvas.drawCircle(center, 2, linePainter);
    var points = _drawPentagon(center, paneSize, canvas, linePainter);
    var step = paneSize/6;
    for(int i = 1; i < 6; i++) {
      _drawPentagon(center, i*step, canvas, linePainter);
    }
    //_drawLabels(center, paneSize, canvas, linePainter);
    canvas.drawCircle(center, 2, linePainter);
    points.forEach((point) => canvas.drawLine(point, center, linePainter));

    // draw values
    valuesList.forEach((el) =>  _drawValues(center, size, paneSize, el, canvas));
  }

  _drawValues(Offset center, Size size, double paneSize, SpiderChartValue model, Canvas canvas) {
    int i = 0;
    var previousPoint;
    canvas.save();
    List<Offset> points = new List();
    var activeStrokePainter = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = 8
      ..color = model?.strokeColor  ?? activeBgColor
      ..isAntiAlias = true;
    var path = Path();
    do {
      assert(model.values[i] <= 1.0 && model.values[i] >= 0, "value must between 0 and 1");
      var value = model.values[i] * paneSize;
      var dx = center.dx + value * cos(i * 2 * pi / nbSides);
      var dy = center.dy + value * sin(i * 2 * pi / nbSides);
      if(previousPoint != null) {
        path.lineTo(dx, dy);
      } else {
        path.moveTo(dx, dy);
      }
      points.add(Offset(dx, dy));
      previousPoint = Offset(dx, dy);
      i++;
    } while (i < nbSides);
    path.lineTo(points[0].dx, points[0].dy);
    var activePainter;
    if(model.gradient != null) {
      activePainter = Paint()
        ..style = PaintingStyle.fill
        ..isAntiAlias = true
        ..shader = model.gradient.createShader(Rect.fromLTWH(0, 0, size.width, size.height));
    } else {
      activePainter = Paint()
        ..style = PaintingStyle.fill
        ..color = model?.bgColor?.withOpacity(.75)  ?? activeBgColor.withOpacity(.75)
        ..isAntiAlias = true;
    }
    canvas.drawPath(path, activePainter);
    canvas.drawPath(path, activeStrokePainter);
    canvas.restore();
  }

  List<Offset> _drawPentagon(Offset center, double paneSize, Canvas canvas, Paint bgPainter) {
    int i = 0;
    var previousPoint;
    canvas.save();
    List<Offset> points = new List();
    var path = Path();
    do {
      var dx = center.dx + paneSize * cos(i * 2 * pi / nbSides);
      var dy = center.dy + paneSize * sin(i * 2 * pi / nbSides);
      if(previousPoint != null) {
        path.lineTo(dx, dy);
      } else {
        path.moveTo(dx, dy);
      }
      points.add(Offset(dx, dy));
      previousPoint = Offset(dx, dy);
      i++;
    } while (i <= nbSides);
    canvas.drawPath(path, bgPainter);
    canvas.restore();
    return points;
  }

  @override
  bool shouldRepaint(SpiderChartDrawer oldDelegate) {
    return false;
  }

  @override
  bool hitTest(Offset position) {
    return false;
  }


}

lib/label_painter.dart
import 'dart:math';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';
import 'offset_extension.dart';

typedef PositionListener(Offset offset);

class SpiderChartLabelDrawer extends CustomPainter {

  final int nbSides, sideIndex;
  final double roundSize;
  final String label;
  final Color lineColor;
  final PositionListener positionListener;
  Offset labelCenter;


  SpiderChartLabelDrawer({
    this.label,
    @required this.nbSides,
    @required this.sideIndex,
    @required this.roundSize,
    @required this.lineColor,
    this.positionListener
  });

  @override
  void paint(Canvas canvas, Size size) {
    var linePainter = Paint()
      ..style = PaintingStyle.stroke
      ..color = this.lineColor
      ..strokeWidth = 2
      ..isAntiAlias = true;
    var paneSize = size.width / 2 - 32;
    var center = Offset(size.width / 2, size.height / 2);
    labelCenter = center.rotate(paneSize, nbSides, sideIndex);
    canvas.drawCircle(labelCenter, roundSize, linePainter);
    if(this.positionListener != null) {
      this.positionListener(labelCenter);
    }
    if(label != null) {
      var labelPainterCenter = center.rotate(paneSize + 4, nbSides, sideIndex);
      var textStyle = TextStyle(color: Colors.white);
      var textSpan = TextSpan(text: label, style: textStyle,);
      var textPainter = TextPainter(text: textSpan, textDirection: TextDirection.ltr, textAlign: TextAlign.center);
      textPainter.layout();
      if(sin(sideIndex * 2 * pi / nbSides) > 0 ) {
        textPainter.paint(canvas, Offset(labelPainterCenter.dx - textPainter.width / 2, labelPainterCenter.dy + 16));
      } else {
        textPainter.paint(canvas, Offset(labelPainterCenter.dx - textPainter.width / 2, labelPainterCenter.dy - 32));
      }
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }

  @override
  bool hitTest(Offset position) {
    if(labelCenter == null)
      return false;
    var distance = (position - labelCenter).distanceSquared;
    if(distance <= 256) {
      return true;
    }
    return false;
  }

}

lib/pointer_widget.dart
import 'package:flutter/material.dart';
import 'offset_extension.dart';

class PointerWidget extends StatelessWidget {

  final int nbSides, sideIndex;
  final double radius;
  final Color color;
  final Gradient gradient;


  PointerWidget({
    this.nbSides, this.sideIndex, this.radius,
    this.color,
    this.gradient
  });

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: PointerDrawer(
        sideIndex: sideIndex,
        nbSides: nbSides,
        radius: radius,
        color: color,
        gradient: gradient
      ),
    );
  }

}



class PointerDrawer extends CustomPainter {

  final int nbSides, sideIndex;
  double radius;
  Color color;
  Gradient gradient;

  PointerDrawer({
    @required this.radius,
    @required this.color,
    this.gradient,
    this.nbSides,
    this.sideIndex
  });

  @override
  void paint(Canvas canvas, Size size) {
    var linePainter = Paint()
      ..style = PaintingStyle.stroke
      ..color = this.color
      ..strokeWidth = 2
      ..isAntiAlias = true;

    var paneSize = size.width / 2 - 32;
    var center = Offset(size.width / 2, size.height / 2);
    var labelCenter = center.rotate(paneSize, nbSides, sideIndex);
    canvas.translate(labelCenter.dx, labelCenter.dy);
    var gradientPainter = Paint()
      ..style = PaintingStyle.fill
      ..isAntiAlias = true
      ..shader = gradient.createShader(Rect.fromCircle(center: Offset(0,0), radius: radius));
    canvas.drawCircle(Offset(0,0), radius, gradientPainter);
    canvas.drawCircle(Offset(0,0), radius, linePainter);
  }

  @override
  bool shouldRepaint(PointerDrawer oldDelegate) {
    return oldDelegate.radius != radius;
  }

  @override
  bool hitTest(Offset position) {
    return false;
  }

}

example/lib/main.dart
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:awesome_spiderchart/chart_painter.dart';
import 'package:awesome_spiderchart/spiderchart.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Spider radar chart demo',
      theme: ThemeData(
        primarySwatch: Colors.blueGrey,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {

  AnimationController _controller, _labelTouchController, _bottomBarController;
  Animation<double> _scaleAnim;
  Animation<double> _scaleChart;
  Animation<double> _scaleLabelAnim;
  Animation<double> _scaleBottomBar;
  int _activeIndex = -1;

  final Gradient bgGradient = new LinearGradient(
    colors: <Color>[
      Colors.blueGrey[500],
      Colors.blueGrey[800],
    ],
    begin: Alignment(1.0, -1.0),
    end: Alignment(-1.0, 1.0),
  );

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(vsync: this, duration: new Duration(milliseconds: 3000));
    _labelTouchController = AnimationController(vsync: this, duration: Duration(milliseconds: 1500));
    _bottomBarController = AnimationController(vsync: this, duration: Duration(milliseconds: 500));

    _scaleLabelAnim = new CurvedAnimation(parent: _labelTouchController, curve: Curves.ease);
    _scaleBottomBar = new CurvedAnimation(parent: _bottomBarController, curve: Curves.ease);
    _scaleAnim = new CurvedAnimation(parent: _controller, curve: Interval(0, .8, curve: Curves.elasticOut));
    _scaleChart = new CurvedAnimation(parent: _controller, curve: Interval(.7, 1, curve: Curves.easeInOutExpo));
    //_scaleLabelAnim.addListener(() => this.setState(() {}));

    Future.delayed(Duration(seconds: 1), () => _controller.forward());
  }

  @override
  Widget build(BuildContext context) {
    final Gradient gradient = new LinearGradient(
      colors: <Color>[
        Colors.lightGreenAccent[100].withOpacity(0.7),
        Colors.lightGreenAccent[700].withOpacity(0.7),
      ],
      begin: Alignment(1.0, -1.0),
      end: Alignment(-1.0, 1.0),
    );
    //---------------------------------------------------
    return Scaffold(
      appBar: AppBar(
        title: Text("Apparence.io - spider chart"),
      ),
      body: Container(
        decoration: BoxDecoration(
          gradient: bgGradient
//            color: Colors.blueGrey[800]
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Expanded(
              child: AnimatedBuilder(
                animation: _scaleAnim,
                builder: (context, child) =>
                  Transform.rotate(
                    angle: -2 * pi * _scaleAnim.value,
                    child: Transform.scale(
                      scale: _scaleAnim.value,
                      child: _buildSpiderChartWidget(gradient),
                    ),
                  ),
                child: _buildSpiderChartWidget(gradient),
              ),
            ),
            Flexible(
              child: Padding(
                padding: const EdgeInsets.only(top: 64.0),
                child: Text(""),
              ),
            )
          ],
        ),
      ),
      bottomSheet: _buildDetailsExpandableCard(context),
    );
  }

  _buildSpiderChartWidget(Gradient gradient) {
    return SpiderChartWidget(
      lineColor: Colors.blueGrey[500],
      activeLabelColor: Colors.blueGrey[50],
      nbSides: 5,
      labels: [
        SpiderChartStatModel(roundSize: 4, label: "test 1"),
        SpiderChartStatModel(roundSize: 4, label: "test 2"),
        SpiderChartStatModel(roundSize: 4, label: "test 3"),
        SpiderChartStatModel(roundSize: 4, label: "test 4"),
        SpiderChartStatModel(roundSize: 4, label: "test 5"),
      ],
      activeLabel: _activeIndex,
      onTap: (index) {
        setState(() {
          _activeIndex = index;
        });
        _bottomBarController.reverse()
          .then((_) => _bottomBarController.forward());
      },
      valuesList: new List()
        ..add(new SpiderChartValue(
          [
            .5 * _scaleChart.value,
            .8 * _scaleChart.value,
            .9 * _scaleChart.value,
            .8 * _scaleChart.value,
            .9 * _scaleChart.value,
            .4 * _scaleChart.value,
            .4 * _scaleChart.value,
            .6 * _scaleChart.value,
          ],
          gradient: gradient,
          strokeColor: Colors.lightGreenAccent[200])
        ),
    );
  }

  _buildDetailsExpandableCard(BuildContext context) {
    return AnimatedBuilder(
      animation: _scaleBottomBar,
      builder: (context, child) => Container(
        height: _scaleBottomBar.value * 220,
        decoration: BoxDecoration(
//          borderRadius: BorderRadius.vertical(top: Radius.circular(32)),
          color: Colors.blueGrey[200]
        ),
        child: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Flexible(
                child: Text("Lorem ipsum lorem ipsum",
                  style: TextStyle(
                    color: Colors.blueGrey[800],
                    fontWeight: FontWeight.bold,
                    fontSize: 21
                  ),
                ),
              ),
              Flexible(
                child: Text("Lorem ipsum lorem ipsum lorem ipsum lorem ipsum " +
                  "lorem ipsum lorem ipsum lorem ipsum lorem ipsum " +
                  "lorem ipsum lorem ipsum lorem ipsum lorem ipsum",
                  style: TextStyle(
                    color: Colors.blueGrey[800],
                    fontWeight: FontWeight.normal,
                    fontSize: 16
                  )
                ),
              ),
              Flexible(
                flex: 1,
                child: Padding(
                  padding: const EdgeInsets.only(top: 24.0),
                  child: Container(
                    decoration: BoxDecoration(
                      color: Colors.transparent,
                      border: Border.all(color: Colors.white, width: 1),
                      borderRadius: BorderRadius.all(Radius.circular(32))
                    ),
                    child: FlatButton(
                      onPressed: () => _bottomBarController.reverse(),
                      child: Text("Close", style: TextStyle(color: Colors.white),)
                    ),
                  ),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }


}
